pkg/register/component.go

```diff
@@ -0,0 +1,227 @@
+type KeycloakClient interface {
+	CreateUser(accessToken string, realmName string, targetRealmName string, user kc.UserRepresentation) (string, error)
+	GetUser(accessToken string, realmName, userID string) (kc.UserRepresentation, error)
+	DeleteUser(accessToken string, realmName, userID string) error
+	ExecuteActionsEmail(accessToken string, realmName string, userID string, actions []string, paramKV ...string) error
+}
+
+// ConfigurationDBModule is the interface of the configuration module.
+type ConfigurationDBModule interface {
+	GetConfiguration(context.Context, string) (dto.RealmConfiguration, error)
+}
+
+// Component is the register component interface.
+type Component interface {
+	RegisterUser(ctx context.Context, realmName string, user apiregister.User) (string, error)
+}
+
+// Component is the management component.
+type component struct {
+	keycloakClient KeycloakClient
+	tokenProvider  keycloak.OidcTokenProvider
+	usersDBModule  UsersDBModule
+	configDBModule ConfigurationDBModule
+	eventsDBModule database.EventsDBModule
+	logger         internal.Logger
+}
+
+// NewComponent returns the management component.
+func NewComponent(keycloakClient KeycloakClient, tokenProvider keycloak.OidcTokenProvider, usersDBModule UsersDBModule, configDBModule ConfigurationDBModule, eventsDBModule database.EventsDBModule, logger internal.Logger) Component {
+	return &component{
+		keycloakClient: keycloakClient,
+		tokenProvider:  tokenProvider,
+		usersDBModule:  usersDBModule,
+		configDBModule: configDBModule,
+		eventsDBModule: eventsDBModule,
+		logger:         logger,
+	}
+}
+
+func (c *component) reportEvent(ctx context.Context, apiCall string, values ...string) {
+	errEvent := c.eventsDBModule.ReportEvent(ctx, apiCall, "back-office", values...)
+	if errEvent != nil {
+		//store in the logs also the event that failed to be stored in the DB
+		internal.LogUnrecordedEvent(ctx, c.logger, apiCall, errEvent.Error(), values...)
+	}
+}
+
+func (c *component) RegisterUser(ctx context.Context, realmName string, user apiregister.User) (string, error) {
+	// Validate input request
+	var err = user.Validate()
+	if err != nil {
+		c.logger.Info(ctx, "err", err.Error())
+		return "", err
+	}
+
+	// Get Realm configuration from database
+	var realmConf dto.RealmConfiguration
+	realmConf, err = c.configDBModule.GetConfiguration(ctx, realmName)
+	if err != nil {
+		c.logger.Info(ctx, "msg", "Can't get realm configuration from database", "err", err.Error())
+		return "", err
+	}
+
+	// Get an OIDC token to be able to request Keycloak
+	var accessToken string
+	accessToken, err = c.tokenProvider.ProvideToken(ctx)
+	if err != nil {
+		c.logger.Info(ctx, "msg", "Can't get OIDC token", "err", err.Error())
+		return "", err
+	}
+
+	err = c.checkExistingUser(ctx, accessToken, realmName, user)
+	if err != nil {
+		return "", err
+	}
+
+	// Generate an unused
+	var username string
+	username, err = c.generateUnusedUsername(ctx, realmName)
```

Comment:
As we first generate an unused username and then create the user, it is not atomic so in very very unlikely situation we may have an issue. 
What about generate a username, and then try to create user. If there is conflict, try again with new username. We can have a loop of 10 tries for ex. 
What do you think ?